app/delivery/send_to_providers.py

```diff
@@ -128,8 +135,32 @@ def check_file_url(file_info: Dict[str, str], notification_id: UUID):
         raise InvalidUrlException
 
 
+def check_for_malware_errors(document_download_response_code, notification):
+    """
+    Check verdict and download calls to the document-download-api will
+    return error codes if the scan is in progress or if malware was detected.
+    This function contains the logic for handling these errors.
+    """
+    # 423 "Locked" response is sent if malicious content was detected
+    if document_download_response_code == 423:
+        current_app.logger.info(
+            f"Malicious content detected! Download and attachment failed for notification.id: {notification.id}"
+        )
+        # Update notification that it contains malware
+        malware_failure(notification=notification)
+
+    # 428 "Precondition Required" response is sent if the scan is still in progress
+    if document_download_response_code == 428:
+        current_app.logger.info(f"Malware scan in progress, could not download files for notification.id: {notification.id}")
+        # Throw error so celery will retry in sixty seconds
+        raise MalwareScanInProgressException
+
+    if document_download_response_code != 200:
+        document_download_internal_error(notification=notification)
```

Comment:
Based on parent's logic flow, this would quickly set a failure on the notification status in the database, without retry strategies. Let's talk if I understand this correctly and if yes, why we might want to give retries a try (pun not intended) for resiliency purpose.

Reply:
Yes you are understanding that correctly. I intended this error to happen when the dd-api has a `DocumentStoreError` which results in a 404: https://github.com/cds-snc/notification-document-download-api/blob/main/app/download/views.py#L131-L139

That happens when there is a `BotoClientError`: https://github.com/cds-snc/notification-document-download-api/blob/dce41bbcc5090d07552aab14677e56c34f036864/app/utils/store.py#L126

My guess is that should only happen if the object is not in S3, in which case I don't think we will recover from the error. But I don't know that for sure.

So I think we could either:
a) target this just to 404 errors, or 
b) let the retry logic happen for this error.

What do you think?

Reply:
I guess we don't really know under what circumstances this error will happen. I will change the error handling so this gets retried as per Jimmy's suggestion since that is safer.